README.md
[DateTime]: https://api.dart.dev/stable/2.8.4/dart-core/DateTime-class.html
[DateFormat]: https://pub.dev/documentation/intl/latest/intl/DateFormat-class.html

# Indonesian Formatter for Dart
This is a library for Indonesian formatting, provides [DateTime][] and number formatting.

# Waktu class
Translate to Indonesian for weekday names, month names, and quarter names, inspired by [DateFormat][].

## Example
```dart
import 'package:format_indonesia/format_indonesia.dart';

void main() {
  // using current time
  var waktu = Waktu();
  print(waktu.yMMMMEEEEd()); // Jumat, 19 Juni 2020 (this was the current time when development)

  // using specific time
  DateTime datetime = DateTime.parse('2020-05-31 12:34:13');
  print(Waktu(datetime).yMMMMEEEEd()); // Minggu, 31 Mei 2020
}
```

## Supported formatters
```
Formatter                       Result
---------                       ------
Waktu().E()                     Jum
Waktu().EEEE()                  Jumat
Waktu().LLL()                   Jun
Waktu().LLLL()                  Juni
Waktu().MMM()                   Jun
Waktu().MMMd()                  19 Jun
Waktu().MMMEd()                 Jum, 19 Jun
Waktu().MMMM()                  Juni
Waktu().MMMMd()                 19 Juni
Waktu().MMMMEEEEd()             Jumat, 19 Juni
Waktu().QQQQ()                  Kuartal kedua
Waktu().yMd()                   19/6/2020
Waktu().yMEd()                  Jum, 19/6/2020
Waktu().yMMM()                  Jun 2020
Waktu().yMMMd()                 19 Jun 2020
Waktu().yMMMEd()                Jum, 19 Jun 2020
Waktu().yMMMM()                 Juni 2020
Waktu().yMMMMd()                19 Juni 2020
Waktu().yMMMMEEEEd()            Jumat, 19 Juni 2020
```

## Pattern formatter
```dart
import 'package:format_indonesia/format_indonesia.dart';

void main() {
  // using current time
  var waktu = Waktu();
  print(waktu.format('dd MMM yyyy')); // 19 Jun 2020 (this was the current time when development)

  // using specific time
  DateTime datetime = DateTime.parse('2020-05-31 12:34:13');
  print(Waktu(datetime).format('EEEE, d MMMM y')); // Minggu, 31 Mei 2020
}
```
Supported patterns is referenced to [DateFormat].

# Numerik class
In this version, provides number to text convertion.

## Example
```dart
import 'package:format_indonesia/format_indonesia.dart';

void main() {
  var numerik = Numerik(911);
  print(numerik.terbilang()); // sembilan ratus sebelas
}
```

# Thank you.
It is really helpful if there are ideas, suggestions, or input for next improvement.


CHANGELOG.md
## 0.1.0
* First release
* Waktu class
* Numerik class (convert number < 1000000000000000 to text)


test/waktu_test.dart
import 'package:format_indonesia/format_indonesia.dart';
import 'package:test/test.dart';

void main() {
  String regexHariPendek = 'Sen|Sel|Rab|Kam|Jum|Sab|Min';
  String regexHari = 'Senin|Selasa|Rabu|Kamis|Jumat|Sabtu|Minggu';
  String regexBulanPendek = 'Jan|Feb|Mar|Apr|Mei|Jun|Jul|Agu|Sep|Okt|Nov|Des';
  String regexBulan = 'Januari|Februari|Maret|April|Mei|Juni|Juli|Agustus|September|Oktober|November|Desember';
  String regexKuartal = 'pertama|kedua|ketiga|keempat';

  test('Print E', () {
    String waktu = Waktu().E();
    print(waktu);
    expect(waktu, matches(regexHariPendek));
  });

  test('Print EEEE', () {
    String waktu = Waktu().EEEE();
    print(waktu);
    expect(waktu, matches(regexHari));
  });

  test('Print LLL', () {
    String waktu = Waktu().LLL();
    print(waktu);
    expect(waktu, matches(regexBulanPendek));
  });

  test('Print LLLL', () {
    String waktu = Waktu().LLLL();
    print(waktu);
    expect(waktu, matches(regexBulan));
  });

  test('Print MMM', () {
    String waktu = Waktu().MMM();
    print(waktu);
    expect(waktu, matches(regexBulanPendek));
  });

  test('Print MMMd', () {
    String waktu = Waktu().MMMd();
    print(waktu);
    expect(waktu, matches('\\d{1,2}\\s(' + regexBulanPendek + ')'));
  });

  test('Print MMMEd', () {
    String waktu = Waktu().MMMEd();
    print(waktu);
    expect(waktu, matches('(' + regexHariPendek + '),\\s\\d{1,2}\\s(' + regexBulanPendek + ')'));
  });

  test('Print MMMM', () {
    String waktu = Waktu().MMMM();
    print(waktu);
    expect(waktu, matches(regexBulan));
  });

  test('Print MMMMd', () {
    String waktu = Waktu().MMMMd();
    print(waktu);
    expect(waktu, matches('\\d{1,2}\\s(' + regexBulan + ')'));
  });

  test('Print MMMMEEEEd', () {
    String waktu = Waktu().MMMMEEEEd();
    print(waktu);
    expect(waktu, matches('(' + regexHari + '),\\s\\d{1,2}\\s(' + regexBulan + ')'));
  });

  test('Print QQQQ', () {
    String waktu = Waktu().QQQQ();
    print(waktu);
    expect(waktu, matches('Kuartal\\s(' + regexKuartal + ')'));
  });

  test('Print yMd', () {
    String waktu = Waktu().yMd();
    print(waktu);
    expect(waktu, matches('\\d{1,2}/\\d{1,2}/\\d{1,4}'));
  });

  test('Print yMEd', () {
    String waktu = Waktu().yMEd();
    print(waktu);
    expect(waktu, matches('(' + regexHariPendek + '),\\s\\d{1,2}/\\d{1,2}/\\d{1,4}'));
  });

  test('Print yMMM', () {
    String waktu = Waktu().yMMM();
    print(waktu);
    expect(waktu, matches('(' + regexBulanPendek + ')\\s\\d{1,4}'));
  });

  test('Print yMMMd', () {
    String waktu = Waktu().yMMMd();
    print(waktu);
    expect(waktu, matches('\\d{1,2}\\s(' + regexBulanPendek + ')\\s\\d{1,4}'));
  });

  test('Print yMMMEd', () {
    String waktu = Waktu().yMMMEd();
    print(waktu);
    expect(waktu, matches('(' + regexHariPendek + '),\\s\\d{1,2}\\s(' + regexBulanPendek + ')\\s\\d{1,4}'));
  });

  test('Print yMMMM', () {
    String waktu = Waktu().yMMMM();
    print(waktu);
    expect(waktu, matches('(' + regexBulan + ')\\s\\d{1,4}'));
  });

  test('Print yMMMMd', () {
    String waktu = Waktu().yMMMMd();
    print(waktu);
    expect(waktu, matches('\\d{1,2}\\s(' + regexBulan + ')\\s\\d{1,4}'));
  });

  test('Print yMMMMEEEEd', () {
    String waktu = Waktu().yMMMMEEEEd();
    print(waktu);
    expect(waktu, matches('(' + regexHari + '),\\s\\d{1,2}\\s(' + regexBulan + ')\\s\\d{1,4}'));
  });
  
  test('Print EEEE, d MMMM y', () {
    String waktu = Waktu().format('EEEE, d MMMM y');
    print(waktu);
    expect(waktu, matches('(' + regexHari + '),\\s\\d{1,2}\\s(' + regexBulan + ')\\s\\d{1,4}'));
  });

  test('Print dd MMM yyyy', () {
    String waktu = Waktu().format('dd MMM yyyy');
    print(waktu);
    expect(waktu, matches('\\d{1,2}\\s(' + regexBulanPendek + ')\\s\\d{1,4}'));
  });
}


test/numerik_test.dart
import 'package:format_indonesia/format_indonesia.dart';
import 'package:test/test.dart';

void main() {
  // Exception: Not accept negative number
  // test('Terbilang -1', () {
  //   String terbilang = Numerik(-1).terbilang();
  //   print(terbilang);
  //   expect(terbilang, equals(''));
  // });

  test('Terbilang 0', () {
    String terbilang = Numerik(0).terbilang();
    print(terbilang);
    expect(terbilang, equals('nol'));
  });

  test('Terbilang 1', () {
    String terbilang = Numerik(1).terbilang();
    print(terbilang);
    expect(terbilang, equals('satu'));
  });

  test('Terbilang 5', () {
    String terbilang = Numerik(5).terbilang();
    print(terbilang);
    expect(terbilang, equals('lima'));
  });

  test('Terbilang 10', () {
    String terbilang = Numerik(10).terbilang();
    print(terbilang);
    expect(terbilang, equals('sepuluh'));
  });

  test('Terbilang 11', () {
    String terbilang = Numerik(11).terbilang();
    print(terbilang);
    expect(terbilang, equals('sebelas'));
  });

  test('Terbilang 15', () {
    String terbilang = Numerik(15).terbilang();
    print(terbilang);
    expect(terbilang, equals('lima belas'));
  });

  test('Terbilang 20', () {
    String terbilang = Numerik(20).terbilang();
    print(terbilang);
    expect(terbilang, equals('dua puluh'));
  });

  test('Terbilang 89', () {
    String terbilang = Numerik(89).terbilang();
    print(terbilang);
    expect(terbilang, equals('delapan puluh sembilan'));
  });

  test('Terbilang 100', () {
    String terbilang = Numerik(100).terbilang();
    print(terbilang);
    expect(terbilang, equals('seratus'));
  });

  test('Terbilang 105', () {
    String terbilang = Numerik(105).terbilang();
    print(terbilang);
    expect(terbilang, equals('seratus lima'));
  });

  test('Terbilang 110', () {
    String terbilang = Numerik(110).terbilang();
    print(terbilang);
    expect(terbilang, equals('seratus sepuluh'));
  });

  test('Terbilang 111', () {
    String terbilang = Numerik(111).terbilang();
    print(terbilang);
    expect(terbilang, equals('seratus sebelas'));
  });

  test('Terbilang 115', () {
    String terbilang = Numerik(115).terbilang();
    print(terbilang);
    expect(terbilang, equals('seratus lima belas'));
  });

  test('Terbilang 150', () {
    String terbilang = Numerik(150).terbilang();
    print(terbilang);
    expect(terbilang, equals('seratus lima puluh'));
  });

  test('Terbilang 500', () {
    String terbilang = Numerik(500).terbilang();
    print(terbilang);
    expect(terbilang, equals('lima ratus'));
  });

  test('Terbilang 555', () {
    String terbilang = Numerik(555).terbilang();
    print(terbilang);
    expect(terbilang, equals('lima ratus lima puluh lima'));
  });

  test('Terbilang 911', () {
    String terbilang = Numerik(911).terbilang();
    print(terbilang);
    expect(terbilang, equals('sembilan ratus sebelas'));
  });

  test('Terbilang 1000', () {
    String terbilang = Numerik(1000).terbilang();
    print(terbilang);
    expect(terbilang, equals('seribu'));
  });

  test('Terbilang 1101', () {
    String terbilang = Numerik(1101).terbilang();
    print(terbilang);
    expect(terbilang, equals('seribu seratus satu'));
  });

  test('Terbilang 111.010', () {
    String terbilang = Numerik(111010).terbilang();
    print(terbilang);
    expect(terbilang, equals('seratus sebelas ribu sepuluh'));
  });

  test('Terbilang 1.050.005', () {
    String terbilang = Numerik(1050005).terbilang();
    print(terbilang);
    expect(terbilang, equals('satu juta lima puluh ribu lima'));
  });

  test('Terbilang 501.501.501', () {
    String terbilang = Numerik(501501501).terbilang();
    print(terbilang);
    expect(terbilang, equals('lima ratus satu juta lima ratus satu ribu lima ratus satu'));
  });

  test('Terbilang 100.115.100.115', () {
    String terbilang = Numerik(100115100115).terbilang();
    print(terbilang);
    expect(terbilang, equals('seratus milyar seratus lima belas juta seratus ribu seratus lima belas'));
  });

  test('Terbilang 510.000.510.000.510', () {
    String terbilang = Numerik(510000510000510).terbilang();
    print(terbilang);
    expect(terbilang, equals('lima ratus sepuluh triliun lima ratus sepuluh juta lima ratus sepuluh'));
  });

  // Exception: Out of limit convertion
  // test('Terbilang 150.051.150.051.150.051', () {
  //   String terbilang = Numerik(150051150051150051).terbilang();
  //   print(terbilang);
  //   expect(terbilang, equals(''));
  // });
}


lib/format_indonesia.dart
import 'package:intl/intl.dart';

class Waktu {
  
  DateTime dateTime;

  static const List<String> _hari = ['Senin', 'Selasa', 'Rabu', 'Kamis', 'Jumat', 'Sabtu', 'Minggu'];
  static const List<String> _bulan = ['Januari', 'Februari', 'Maret', 'April', 'Mei', 'Juni', 'Juli', 'Agustus', 'September', 'Oktober', 'November', 'Desember'];
  static const List<String> _kuartal = ['pertama', 'kedua', 'ketiga', 'keempat'];

  Waktu([DateTime dateTime]) {
    this.dateTime = dateTime ?? DateTime.now();
  }

  String E() => _hari[this.dateTime.weekday - 1].substring(0, 3);
  String EEEE() => _hari[this.dateTime.weekday - 1];
  String LLL() => _bulan[this.dateTime.month - 1].substring(0, 3);
  String LLLL() => _bulan[this.dateTime.month - 1];
  String MMM() => _bulan[this.dateTime.month - 1].substring(0, 3);
  String MMMd() => '${this.dateTime.day.toString()} ${_bulan[this.dateTime.month - 1].substring(0, 3)}';
  String MMMEd() => '${_hari[this.dateTime.weekday - 1].substring(0, 3)}, ${this.dateTime.day.toString()} ${_bulan[this.dateTime.month - 1].substring(0, 3)}';
  String MMMM() => _bulan[this.dateTime.month - 1];
  String MMMMd() => '${this.dateTime.day.toString()} ${_bulan[this.dateTime.month - 1]}';
  String MMMMEEEEd() => '${_hari[this.dateTime.weekday - 1]}, ${this.dateTime.day.toString()} ${_bulan[this.dateTime.month - 1]}';
  String QQQQ() => 'Kuartal ${_kuartal[((this.dateTime.month - 1) / 3).floor()]}';
  String yMd() => '${this.dateTime.day.toString()}/${this.dateTime.month.toString()}/${this.dateTime.year.toString()}';
  String yMEd() => '${_hari[this.dateTime.weekday - 1].substring(0, 3)}, ${this.dateTime.day.toString()}/${this.dateTime.month.toString()}/${this.dateTime.year.toString()}';
  String yMMM() => '${_bulan[this.dateTime.month - 1].substring(0, 3)} ${this.dateTime.year.toString()}';
  String yMMMd() => '${this.dateTime.day.toString()} ${_bulan[this.dateTime.month - 1].substring(0, 3)} ${this.dateTime.year.toString()}';
  String yMMMEd() => '${_hari[this.dateTime.weekday - 1].substring(0, 3)}, ${this.dateTime.day.toString()} ${_bulan[this.dateTime.month - 1].substring(0, 3)} ${this.dateTime.year.toString()}';
  String yMMMM() => '${_bulan[this.dateTime.month - 1]} ${this.dateTime.year.toString()}';
  String yMMMMd() => '${this.dateTime.day.toString()} ${_bulan[this.dateTime.month - 1]} ${this.dateTime.year.toString()}';
  String yMMMMEEEEd() => '${_hari[this.dateTime.weekday - 1]}, ${this.dateTime.day.toString()} ${_bulan[this.dateTime.month - 1]} ${this.dateTime.year.toString()}';

  String format(String format) {
    format = format.replaceAll('EEEE', "'${this.EEEE()}'");
    format = format.replaceAll('E', "'${this.E()}'");
    format = format.replaceAll('LLLL', "'${this.LLLL()}'");
    format = format.replaceAll('LLL', "'${this.LLL()}'");
    format = format.replaceAll('MMMM', "'${this.MMMM()}'");
    format = format.replaceAll('MMM', "'${this.MMM()}'");

    return DateFormat(format).format(this.dateTime);
  }
}


class Numerik {

  int number;

  static const List<String> _bilang = ['', 'satu', 'dua', 'tiga', 'empat', 'lima', 'enam', 'tujuh', 'delapan', 'sembilan'];

  Numerik(this.number);

  String _doTerbilang(int number) {
    if (number < 10) return _bilang[number];
    if (number < 20) {
      int modsepuluh = number % 10;
      if (modsepuluh == 0) return 'sepuluh';
      else if (modsepuluh == 1) return 'sebelas';
      return '${_bilang[modsepuluh]} belas';
    }
    if (number < 100) {
      int divsepuluh = number ~/ 10;
      int modsepuluh = number % 10;
      String result = '${_bilang[divsepuluh]} puluh';
      if (modsepuluh > 0) result = '$result ${_bilang[modsepuluh]}';
      return result;
    }
    if (number < 1000) {
      int divseratus = number ~/ 100;
      int modseratus = number % 100;
      String result = '';
      if (divseratus == 1) result = 'seratus';
      else result = '${_bilang[divseratus]} ratus';
      if (modseratus > 0) result = '$result ${_doTerbilang(modseratus)}';
      return result;
    }
    if (number < 1000000) {
      int divseribu = number ~/ 1000;
      int modseribu = number % 1000;
      String result = '';
      if (divseribu == 1) result = 'seribu';
      else result = '${_doTerbilang(divseribu)} ribu';
      if (modseribu > 0) result = '$result ${_doTerbilang(modseribu)}';
      return result;
    }
    if (number < 1000000000) {
      int divjuta = number ~/ 1000000;
      int modjuta = number % 1000000;
      String result = '${_doTerbilang(divjuta)} juta';
      if (modjuta > 0) result = '$result ${_doTerbilang(modjuta)}';
      return result;
    }
    if (number < 1000000000000) {
      int divm = number ~/ 1000000000;
      int modm = number % 1000000000;
      String result = '${_doTerbilang(divm)} milyar';
      if (modm > 0) result = '$result ${_doTerbilang(modm)}';
      return result;
    }
    if (number < 1000000000000000) {
      int divt = number ~/ 1000000000000;
      int modt = number % 1000000000000;
      String result = '${_doTerbilang(divt)} triliun';
      if (modt > 0) result = '$result ${_doTerbilang(modt)}';
      return result;
    }
    return '';
  }

  String terbilang() {
    if (this.number < 0) throw Exception('Not accept negative number');
    if (this.number >= 1000000000000000) throw Exception('Out of limit convertion');
    if (this.number == 0) return 'nol';
    return _doTerbilang(this.number);
  }
}
